chainerrl/agents/trpo.py

```diff
@@ -350,27 +418,94 @@ def _update_policy(self, dataset):
 
         # Recompute action distributions for batch backprop
         action_distrib = self.policy(states)
-        action_distrib_old = action_distrib.copy()
+
+        log_prob_old = xp.array(
+            [transition['log_prob'] for transition in dataset],
+            dtype=np.float32)
 
         gain = self._compute_gain(
-            action_distrib=action_distrib,
-            action_distrib_old=action_distrib_old,
-            actions=actions,
+            log_prob=action_distrib.log_prob(actions),
+            log_prob_old=log_prob_old,
+            entropy=action_distrib.entropy,
             advs=advs)
 
+        # Distribution to compute KL div against
+        action_distrib_old = action_distrib.copy()
+
         full_step = self._compute_kl_constrained_step(
             action_distrib=action_distrib,
             action_distrib_old=action_distrib_old,
             gain=gain)
 
         self._line_search(
             full_step=full_step,
-            states=states,
-            actions=actions,
+            dataset=dataset,
             advs=advs,
             action_distrib_old=action_distrib_old,
             gain=gain)
 
+    def _update_policy_recurrent(self, dataset):
+        """Update the policy using a given dataset.
+
+        The policy is updated via CG and line search.
+        """
+
+        xp = self.model.xp
+        flat_transitions = list(itertools.chain.from_iterable(dataset))
+
+        # Prepare data for a recurrent model
+        seqs_states = []
+        for ep in dataset:
+            states = self.batch_states(
+                [transition['state'] for transition in ep], xp, self.phi)
+            if self.obs_normalizer:
+                states = self.obs_normalizer(states, update=False)
+            seqs_states.append(states)
+
+        flat_actions = xp.array(
+            [transition['action'] for transition in flat_transitions])
+        flat_advs = xp.array(
+            [transition['adv'] for transition in flat_transitions],
+            dtype=np.float32)
+
+        with chainer.using_config('train', False),\
+                chainer.no_backprop_mode():
+            policy_rs = self.policy.concatenate_recurrent_states(
+                [ep[0]['recurrent_state'][0] for ep in dataset])
+
+        flat_distribs, _ = self.policy.n_step_forward(
+            seqs_states, recurrent_state=policy_rs, output_mode='concat')
+
+        if self.standardize_advantages:
+            mean_advs = xp.mean(flat_advs)
+            std_advs = xp.std(flat_advs)
+            flat_advs = (flat_advs - mean_advs) / (std_advs + 1e-8)
```

Comment:
Non-recurrent code standardizes advantages before computing action distributions.  Could you make their order consistent?